src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -4079,33 +4079,70 @@ void C2_MacroAssembler::vector_mask_operation(int opc, Register dst, KRegister m
       bsfq(tmp, tmp);
       cmov(Assembler::notZero, dst, tmp);
       break;
+    case Op_VectorMaskToLong:
+      break;
     default: assert(false, "Unhandled mask operation");
   }
 }
 
 void C2_MacroAssembler::vector_mask_operation(int opc, Register dst, XMMRegister mask, XMMRegister xtmp,
-                                              XMMRegister xtmp1, Register tmp, int masklen, int masksize,
-                                              int vec_enc) {
+                                              int masklen, BasicType bt, int vec_enc) {
   assert(VM_Version::supports_avx(), "");
-  vpxor(xtmp, xtmp, xtmp, vec_enc);
-  vpsubb(xtmp, xtmp, mask, vec_enc);
-  vpmovmskb(tmp, xtmp, vec_enc);
-  if (masksize < 16) {
-    andq(tmp, (((jlong)1 << masklen) - 1));
+  assert(masklen <= 32, "");
+  bool need_clip = false;
+  switch(bt) {
+    case T_BYTE:
+      vpmovmskb(dst, mask, vec_enc);
+      need_clip = masklen < 16;
+      break;
+    case T_SHORT:
+      vpacksswb(xtmp, mask, mask, vec_enc);
+      if (masklen >= 16) {
+        assert(masklen == 16, "");
+        vpermpd(xtmp, xtmp, 8, vec_enc);
+      }
+      vpmovmskb(dst, xtmp, Assembler::AVX_128bit);
+      need_clip = masklen < 16;
+      break;
+    case T_INT:
+    case T_FLOAT:
+      vmovmskps(dst, mask, vec_enc);
+      need_clip = masklen < 4;
+      break;
+    case T_LONG:
+    case T_DOUBLE:
+      vmovmskpd(dst, mask, vec_enc);
+      need_clip = masklen < 2;
+      break;
+    default: assert(false, "Unhandled type, %s", type2name(bt));
   }
   switch(opc) {
     case Op_VectorMaskTrueCount:
-      popcntq(dst, tmp);
+      if (need_clip) {
+        // need_clip implies masklen < 32
+        andl(dst, (1 << masklen) - 1);
+      }
+      popcntl(dst, dst);
       break;
     case Op_VectorMaskLastTrue:
-      mov64(dst, -1);
-      bsrq(tmp, tmp);
-      cmov(Assembler::notZero, dst, tmp);
+      if (need_clip) {
+        // need_clip implies masklen < 32
+        andl(dst, (1 << masklen) - 1);
+      }
+      lzcntl(dst, dst);
+      negl(dst);
+      addl(dst, 31);
```

Comment:
I think from latency perspective earlier sequence was   better given that  constant move to a register is not scheduled to an execution port.